library/std/src/diagnostics.qs

```diff
@@ -149,5 +149,73 @@ namespace Microsoft.Quantum.Diagnostics {
         areEqual
     }
 
-    export DumpMachine, DumpRegister, CheckZero, CheckAllZero, Fact, CheckOperationsAreEqual;
+    /// # Summary
+    /// Starts counting the number of times the given operation is called. The counter is reset to zero.
+    ///
+    /// # Description
+    /// This operation allows you to count the number of times a given operation is called. As part of
+    /// starting teh counting, the counter is reset to zero, which may override existing counts for the same operation.
+    /// Counting is based on the specific specialization of the operation invoked, so `X` and `Adjoint X` are counted separately.
+    /// Likewise `Controlled X`, `CNOT`, and `CX` are independent operations that are counted separately.
+    ///
+    /// # Input
+    /// ## callable
+    /// The operation to be counted.
+    @Config(Unrestricted)
+    operation StartCountingOperation<'In, 'Out>(callable : 'In => 'Out) : Unit {
+        body intrinsic;
+    }
+
+    /// # Summary
+    /// Stops counting the number of times the given operation is called and returns the count. Returns -1
```

Comment:
Same question, if the operation is not counted, why -1 instead of 0?

Reply:
I wanted to differentiate between "You asked me to count this and it was never called" and "You never asked me to count this" which I thought would be helpful. I'm not wedded to it, though.

Reply:
The incorrect "You never asked me to count this" scenario in user's code won't be caught by returning either of these values in most cases - the checks typically check that the value is greater than a threshold (0 or a positive number of allowed calls) to say "you were not supposed to use that many", they almost never check for "wait, you called too few!". I like to think that this scenario will be caught by the user trying this code on an incorrect solution to see it fail, but I can't think of a way to automate it...

At the same time, there is a check that does arithmetic with numbers of calls to something (if you're allowed to use one or the other but not both), and I'm worried that returning a negative value in this case will mess the arithmetic up.

Actually, this feels like getting an element from a map when it's not there - some languages return 0, some throw exceptions. Should we maybe just throw an exception? This will happen when the teacher writes the testing harness, not when the student writes their code, so it's a good way to bring it to their attention without impacting student experience. As long as the student doesn't reset the counter in their code...

Reply:
That makes sense. I'll turn it into a runtime error instead (both for this and qubit tracking), that way when someone starts tracking after it has already started or stops when it hasn't started they get immediate failures that help them know they are doing something wrong. It will also simplify the doc comments :)